packages/core/src/utils.ts

```diff
@@ -259,6 +263,143 @@ export async function* maskStreamTags(
   }
 }
 
+/**
+ * Resolve serialized zod output - This function takes the string output ot the `jsonSchemaToZod` function
+ * and instantiates the zod object correctly.
+ *
+ * @param schema - serialized zod object
+ * @returns resolved zod object
+ */
+export function resolveSerializedZodOutput(schema: string): z.ZodType {
+  // Creates and immediately executes a new function that takes 'z' as a parameter
+  // The function body is a string that returns the serialized zod schema
+  // When executed with the 'z' parameter, it reconstructs the zod schema in the current context
+  return Function('z', `"use strict";return (${schema});`)(z);
+}
+
+/**
+ * Checks if a tool is a Vercel Tool
+ * @param tool - The tool to check
+ * @returns True if the tool is a Vercel Tool, false otherwise
+ */
+export function isVercelTool(tool: any): tool is VercelTool {
+  // Checks if the tool is an object and has a 'parameters' property
+  return typeof tool === 'object' && tool !== null && 'parameters' in tool;
+}
+
+interface ToolOptions {
+  name: string;
+  runId?: string;
+  threadId?: string;
+  resourceId?: string;
+  logger: Logger;
+  description?: string;
+  mastra?: MastraPrimitives;
+  memory?: MastraMemory;
+  agentName?: string;
+}
+
+type ToolToConvert = VercelTool | ToolAction<any, any, any, any>;
+
+interface LogOptions {
+  agentName?: string;
+  toolName: string;
+  tool?: ToolToConvert;
+  type?: 'tool' | 'toolset';
+}
+
+interface LogMessageOptions {
+  start: string;
+  error: string;
+}
+
+function createLogMessageOptions({ agentName, toolName, tool, type }: LogOptions): LogMessageOptions {
+  // If no agent name, use default format
+  if (!agentName) {
+    return {
+      start: `Executing tool ${toolName}`,
+      error: `Failed tool execution`,
+    };
+  }
+
+  const prefix = `[Agent:${agentName}]`;
+  const vercelPrefix = isVercelTool(tool) ? 'Vercel ' : '';
+  const toolType = type === 'toolset' ? 'toolset' : 'tool';
+
+  return {
+    start: `${prefix} - Executing ${vercelPrefix}${toolType} ${toolName}`,
+    error: `${prefix} - Failed ${vercelPrefix}${toolType} execution`,
+  };
+}
+
+function createExecute(tool: ToolToConvert, options: ToolOptions, logType?: 'tool' | 'toolset') {
+  const { logger, ...rest } = options;
+
+  const { start, error } = createLogMessageOptions({
+    agentName: options.agentName,
+    toolName: options.name,
+    tool,
+    type: logType,
+  });
+
+  const execFunction = async (args: any, execOptions: any) => {
+    if (isVercelTool(tool)) {
+      return tool?.execute?.(args, execOptions) ?? undefined;
+    }
+    return (
+      tool?.execute?.(
+        {
+          context: args,
+          threadId: options.threadId,
+          resourceId: options.resourceId,
+          mastra: options.mastra,
+          memory: options.memory,
```

Comment:
We shouldn't pass in memory, it's available from mastra

Reply:
Didn't we make a recent change where we pass in memory alongside mastra? @TylerBarnes do you have more info on why we did that?

Reply:
I started by passing in the agent and then de-scoped it to just the agent memory to try to reduce api surface